### ISSUE TITLE
Memory usage for large datasets

### ISSUE THREAD

**Issue by guilleva** [randym/axlsx#276]
Hi, I'm facing a memory issue when trying to generate a xlsx file with >600 columns and ~10k rows. The memory usage grows up to 3GB aprox and it's making the dyno be restarted by heroku.

I'm wondering if there is an known way to generate it without making a heavy use of memory.

The way I'm doing it can be represented by this script:

``` ruby
Axlsx::Package.new do |p|
  p.workbook.add_worksheet(:name => "Test") do |sheet|
    11_000.times do
        sheet.add_row ["test data"] * 600
    end
  end
  p.serialize('tmp/test.xlsx')
end
```

**Comment by srpouyet** [comment]
Our Delayed Job workers aren't processing jobs (out of memory), probably because of this issue. Any updates on the status? Cheers!

**Comment by pallymore** [comment]
+1 on this. We are having the same issue with huge datasets :/

**Comment by jeremywadsack** [comment]
We just ended up spinning up more hardware to deal with this for the time being. :/

**Comment by hassanrehman** [comment]
Old thread. We're also having problems. 2.1 MB file needs 650MB of RAM.

Still waiting for a solution

**Comment by jeremywadsack** [comment]
Yeah. We would still love a solution. We run a 24gb server instance so we
can build 15mb files.

- Jeremy

**Comment by jeremywadsack** [comment]
Just had another build crash because a 400MB Excel file took 22GB of memory to build. (Incidentally, Excel used 1.5GB to open this file.)

In the python world we use `xlsxwriter` to generate Excel files which has [a few options to reduce memory](http://xlsxwriter.readthedocs.org/working_with_memory.html) including flushing rows to disk as they are written. There are trade-offs there but it's possible something like that could be done with `axlsx` as well?

**Comment by AaronMegaphone** [comment]
+1 for a solution/update for this.

**Comment by remofritzsche** [comment]
+1 for a solution

**Comment by altjx** [comment]
First of all, thank you for this wonderful gem. I'm also having this problem when trying to generate 30k rows of text. As an alternative (and as opposed to increasing memory), does Axlsx have the capability to append data to an existing spreadsheet or join two of them without exhausting memory?

I'd at least be able to write multiple files and join them together at the end.

**Comment by aalvarado** [comment]
Maybe there could be an option of using a temp file instead of filling a StringIO object, once it is finished it can just place everything in the zip file. Thoughts?

**Comment by jeremywadsack** [comment]
The XLSX file isn't a single file. It's a collection of files that are zipped together. Which I think makes this more complicated.

The python library xlsxwriter has [a feature like this](http://xlsxwriter.readthedocs.org/working_with_memory.html):

> The optimization works by flushing each row after a subsequent row is written. In this way the largest amount of data held in memory for a worksheet is the amount of data required to hold a single row of data.

This does introduce limits in how you can use it:

> Since each new row flushes the previous row, data must be written in sequential row order when 'constant_memory' mode is on:

Perhaps something like this could be implemented in axlsx, though?

**Comment by altjx** [comment]
I haven't taken a look at it just yet, but I'm assuming it's just like a word doc... a bunch of XML files zipped together. However, with word docs you can indeed extract certain contents from the word/document.xml. During the mean time, I'm going to try working on my own solution. Haven't seen much feedback from the guilleva here so not really expecting a fix anytime soon.

**Comment by aalvarado** [comment]
> The XLSX file isn't a single file. It's a collection of files that are zipped together. Which I think makes this more complicated. 

But the file that has the actual rows can it be still be written sequentially no?

looking inside an xlsx file I can see that the `xl/_rel/sharedStrings.xml` has the sheet data in it.

**Comment by pol0nium** [comment]
+1, this is a huge issue for me. I'd be happy to hear about possible workarounds.

**Comment by Wenzel** [comment]
+1, big issue for me too.

**Comment by pol0nium** [comment]
Would it help if I'd add a bounty via [bountysource](https://www.bountysource.com/)? @randym what do you think?

**Comment by chanko** [comment]
+1

**Comment by jeremywadsack** [comment]
#352 is a similar request to append rows to an existing file.

**Comment by Paxa** [comment]
I've been struggling with same issues, when generating > 100k rows our servers started to use swap and become very slow.

Solved by making own excel library https://github.com/paxa/fast_excel, it works 3-4 times faster and consume 5-20 MB ram regardless of data size

**Comment by cseelus** [comment]
The problem with these memory issues we face using AXLSX via the to_spreadsheet gem is, that the memory won't be free after xslx creation. This means we have to manually restart our worker dyno for now and also limit the maximum number of rows per Excel sheet.

@Paxa Nice work. Depending on the answer of the maintainer of the to_spreadsheet gem (allows a Rails app to render Excel files using the existing slim/haml/erb/etc views) I'll see if fast_excel could be used as an alternative to AXLSX.

**Comment by crystianwendel** [comment]
So, @cseelus, i'm having issues with this lib's memory usage too. What did you do? @Paxa 's lib or to_spreadsheet?

Thank you!

**Comment by cseelus** [comment]
For now we limit the number of rows per xslx file to mitigate the memory problems using the to_spreadsheet gem.

**Comment by barttenbrinke** [comment]
According to my memoryprofile skills, the main issue is here: https://github.com/randym/axlsx/blob/master/lib/axlsx/workbook/worksheet/row.rb#L156 This is never released, expanding tot up to 3GB in my testcase.

**Comment by simi** [comment]
@barttenbrinke can you share your code to reproduce and measure this problem?

**Comment by comictvn** [comment]
anyone already fixed the issue ?

**Comment by barttenbrinke** [comment]
@simi https://github.com/SamSaffron/memory_profiler @comictvn This wil not be an easy fix, seeing that row.rb has not been touched in > year, and the fact that this ticket is from 2014, I would not get your hopes up.

**Comment by jeremywadsack** [comment]
Well, yes, @jurriaan said this in 2014: "the fact is that atm axlsx allocates way too much objects and causes the heap to grow very big."

I think there are two approaches here, neither is likely to be "easy":
1. Significantly reduce the size of the `Cell` objects.
2. Flush rows to disk as they are created.

The problem with the first approach is that we might generate a large excel file with 20+ tabs and each of them with five million rows (I know Excel doesn't support that many rows, but the file format does). I think you'd need to reduce the size of the `Cell` object to under 50 bytes to make a significant difference.

The problem with the second approach is that it means the developer would have to write rows in sequential order. That's a change to the interface, so would need to be "opt-in".

**Comment by iamtheschmitzer** [comment]
👍 for solution 2

**Comment by mkhoa1412** [comment]
+1, this is a huge issue for me. I'd be happy to hear about possible workarounds.

**Comment by janklimo** [comment]
We've been very happy with `axlsx` but the memory usage really started to hit us. This may not apply to everyone, but if all you need is simple XLSX formatted output, I'd look into [rubyXL](https://github.com/weshatheleopard/rubyXL). [Here's a benchmark](https://gist.github.com/janklimo/240cf6b4517941f52809343631daff77) showing that its retained memory is constant, while memory retained by `axlsx` grows linearly with the number of rows/columns.

**Comment by gdimitris** [comment]
@randym any update on this ?

**Comment by wangDvp** [comment]
I use the [fast_excel](https://github.com/paxa/fast_excel) instead of this gem,
this is my local benchmark test result
`

    === FastExcel 1000 x 20 ===
    Time: 0.08s
    RSS:  26.3MB -> 28.4MB
    Major GC: +1
    
    === FastExcel 1000 x 50 ===
    Time: 0.16s
    RSS:  28.6MB -> 32.1MB
    Major GC: +1
    
    === FastExcel 1000 x 100 ===
    Time: 0.31s
    RSS:  32.1MB -> 37.9MB
    Major GC: +1
    
    === FastExcel 10000 x 20 ===
    Time: 0.69s
    RSS:  37.9MB -> 51.8MB
    Major GC: +1
    
    === FastExcel 10000 x 50 ===
    Time: 1.64s
    RSS:  51.8MB -> 92.4MB
    Major GC: +1
    
    === FastExcel 10000 x 100 ===
    Time: 3.2s
    RSS:  92.4MB -> 164.4MB
    Major GC: +1
    
    === FastExcel 100000 x 20 ===
    Time: 6.89s
    RSS:  164.4MB -> 318.3MB
    Major GC: +1
    
    === FastExcel 100000 x 50 ===
    Time: 20.41s
    RSS:  318.3MB -> 722.6MB
    Major GC: +1
`
